### PR TITLE
Allow listing a PR's reviews

### PR DESCRIPTION
**Pull request by deiwin**
This allows checking if a PR really is ready for merging, for example. A build
could be failed if there are any PENDING or CHANGES_REQUESTED reviews. Or if
there isn't enough APPROVED reviews.

Also includes small fixes for some other open issues.

### CODE REVIEW

**Comment by deiwin** [comment]
I haven't confirmed that this actually works, though. Don't know how to test changes to this plugin.

File: README.md [hunk 1]
```diff
@@ -93,6 +93,12 @@ pipeline {
 ```
 
 #### Detecting whether a build was started by the trigger in a script:
+
+Note that the following uses `currentBuild.rawBuild` and should therefore only
```
Comment (aaronjwhiteside) [inline README.md]: 👍

File: src/main/java/org/jenkinsci/plugins/pipeline/github/client/Review.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        return state;
+    }
+
+    /**
+     * @param state one of APPROVED, PENDING, CHANGES_REQUESTED, DISMISSED, COMMENTED
+     * @return this review
+     * throws {@link IllegalArgumentException} if state is invalid
+     */
+    public Review setState(final String state) {
+        if (STATE_APPROVED.equals(state) || STATE_PENDING.equals(state) || STATE_CHANGES_REQUESTED.equals(state)
```
Comment (aaronjwhiteside) [inline src/main/java/org/jenkinsci/plugins/pipeline/github/client/Review.java]: You could put the valid states in a Set<> in a static {} block, then check if the provided state is in the Set.

File: src/main/java/org/jenkinsci/plugins/pipeline/github/client/Review.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+public class Review implements Serializable {
+    private static final long serialVersionUID = 1L;
+
+    public static final String STATE_APPROVED = "APPROVED";
+    public static final String STATE_PENDING = "PENDING";
+    public static final String STATE_CHANGES_REQUESTED = "CHANGES_REQUESTED";
+    public static final String STATE_DISMISSED = "DISMISSED";
+    public static final String STATE_COMMENTED  = "COMMENTED";
+
+    private String body;
```
Comment (aaronjwhiteside) [inline src/main/java/org/jenkinsci/plugins/pipeline/github/client/Review.java]: Can you add in `commitId`, here?

https://developer.github.com/v3/pulls/reviews/#list-reviews-on-a-pull-request

As someone pushes more changes to a PR, the new Reviews will be against different commit's, so it would be nice to be able to tell which one the review was against.

File: README.md [hunk 4]
```diff
@@ -491,14 +510,14 @@ for (comment in pullRequest.comments) {
...
-} 
+}
 ```
 
 ### Listing a commit's statuses
 ```groovy
 for (commit in pullRequest.commits) {
   for (status  in commit.statuses) {
-     echo "Commit: ${commit.sha}, Status: ${status.status}, Context: ${status.context}, URL: ${status.targetUrl}"
+     echo "Commit: ${commit.sha}, State: ${status.state}, Context: ${status.context}, URL: ${status.targetUrl}"
```
Comment (aaronjwhiteside) [inline README.md]: 👍

File: README.md [hunk 5]
```diff
@@ -513,7 +532,7 @@ for (commit in pullRequest.commits) {
 ### Listing a Pull Request's current statuses
 ```groovy
 for (status in pullRequest.statuses) {
-  echo "Commit: ${pullRequest.head}, Status: ${status.status}, Context: ${status.context}, URL: ${status.targetUrl}"
+  echo "Commit: ${pullRequest.head}, State: ${status.state}, Context: ${status.context}, URL: ${status.targetUrl}"
```
Comment (aaronjwhiteside) [inline README.md]: 👍

**Comment by aaronjwhiteside** [comment]
It's still a TODO on my list to write automated tests, once I figure out Jenkins's plugin testing framework..

For the moment I've been manually testing things with `mvn hpi:run` and `ngrok` against a test repo on GitHub.. I know this is terrible.. I wouldn't dream of not writing automated tests during my day job :( I'll get around to it sooner or later, happy to have someone hack away at a first attempt though!

File: src/main/java/org/jenkinsci/plugins/pipeline/github/client/Review.java (hunk 3 above)
Comment (deiwin) [reply src/main/java/org/jenkinsci/plugins/pipeline/github/client/Review.java]: https://github.com/jenkinsci/pipeline-github-plugin/pull/18/commits/7fa0d37d8b8edf0eed3027a65075f0e9b019c403

File: src/main/java/org/jenkinsci/plugins/pipeline/github/client/Review.java (hunk 2 above)
Comment (deiwin) [reply src/main/java/org/jenkinsci/plugins/pipeline/github/client/Review.java]: Used a set in https://github.com/jenkinsci/pipeline-github-plugin/pull/18/commits/a84d222694a33f13995ffeb6c490e0270d371c9f.

This was based on [similar implementations in the egit-github library](https://github.com/eclipse/egit-github/blob/v4.11.0.201803080745-r/org.eclipse.egit.github.core/src/org/eclipse/egit/github/core/CommitStatus.java#L151-L163). I stuck with using just String instead of an enum, because of the reasons you mentioned, plus I don't know what the implications are for serialization when using Enums here.

**Comment by deiwin** [comment]
I [tested](https://github.com/salemove/pipeline-github-plugin/pull/1) this manually with `mvn hpi:run` + `ngrok`, as you suggested. It works as expected.

I noticed that the API is a bit peculiar, though. It lists _all_ of the reviews by a single person, not the aggregation (example output visible [in this comment](https://github.com/salemove/pipeline-github-plugin/pull/1#issuecomment-387026842)). So the the users of this plugin will have to build the aggregations themselves, to know if a PR is actually ready for merging or not. Not sure if this should be documented in this plugin. The same is true for commit statuses, which also include multiple states for a single context.

What's even more strange, is that the review statuses don't have a `created_at` field, so one can't use that for knowing which of the reviews was given last. Luckily the reviews seem to come pre-sorted, and their IDs seem to be incrementing, so both of these bits of information could be used for finding the _last review from a person_, but neither of those behaviors is documented and therefore not guaranteed.

File: src/main/java/org/jenkinsci/plugins/pipeline/github/client/Review.java [hunk 6]
```diff
@@ -14,6 +18,10 @@
     public static final String STATE_DISMISSED = "DISMISSED";
     public static final String STATE_COMMENTED = "COMMENTED";
 
+    private static final Set<String> reviewStates = Stream
```
Comment (aaronjwhiteside) [inline src/main/java/org/jenkinsci/plugins/pipeline/github/client/Review.java]: 👍

**Comment by aaronjwhiteside** [comment]
Thanks for putting in the hard work @deiwin, it's much appreciated! 

I'll be cutting a release in the next week or so, so it shouldn't be too long before other can take advantage of your additions.

**Comment by deiwin** [comment]
Any progress on the release, @aaronjwhiteside?

**Comment by deiwin** [comment]
bump

**Comment by aaronjwhiteside** [comment]
Sorry, been a busy few week, will try and do it this weekend!

**Comment by deiwin** [comment]
Thank you!

**Comment by deiwin** [comment]
:(

**Comment by aaronjwhiteside** [comment]
Sorry @deiwin, soon!

**Comment by aaronjwhiteside** [comment]
@deiwin done, version 2.1 released!
